documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+
+# **INFORMATIVE** Integer and Floating Point Types
+While investigating the use of generic math we came to the conclusion that making types which supports both integer and floating point types would not be optimal. This was discussed at length on the discord [here](https://discord.com/channels/521092042781229087/587346162802229298/1167705816812498974). Ultimately it was decided to provide both an integer and floating point variant for each vector type and every type built from them. These types are generic where `Vector2I<T>` will be a 2D vector which takes any binary integer type for `T`. Similarly `Vector2F<T>` will be a 2D vector which takes any floating point type for `T`. By extension we get types like `CubeI<T>` and `RectangleF<T>`. The integer types are granted the bitwise operators `&`, `~`, `|`, and `^`. Floating point types will include some operations that require certain functions unavailable to integer types like `Length` which requires `Sqrt`.
+
+# Vector Types
+
+The main types defined for this proposal are two sets of vector types, `VectorNI<T>` and `VectorNF<T>` where `N` defines the dimensionality of the vector between 2 and 5. Aside from some notable exceptions the vector types will be expected to implement a similar api to the correlating one below (`VectorNI<T>` will follow `Vector2I<T>` and the `VectorNF<T>` will follow `Vector2F<T>`).
+
+Aside from the api below, the following constraints should also be followed:
+- Three Dimensional Vectors should implement a CrossProduct function which takes another vector and returns the cross product with our original vector. A static implementation of this function should be available as well.
+- Every Dimension should have the relevant number of properties (X and Y for Vector2, and X,Y,Z,W,V for Vector5) and relevant unit vectors
+- Constructors for higher dimensions should include lower dimension variants that use the lower dimensions for their specific components (vector2 -> X,Y)
+
+```cs
+public readonly record struct Vector2I<TScalar> : IEquatable<Vector2I<TScalar>>, 
+                           IEquatable<Vector2>,
+                           IBinaryIntegerVector<Vector2I<TScalar>, TScalar>,
+                           IVector2<Vector2I<TScalar>, TScalar>
+                           where TScalar : IBinaryInteger<TScalar>
+{
+    ///<summary>
+    ///The X component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar X { get; }
+
+    ///<summary>
+    ///The Y component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar Y { get; }
+
+    public static Vector2I<TScalar> UnitX { get; }
+    public static Vector2I<TScalar> UnitY { get; }
+
+    public static Vector2I<TScalar> Zero { get; }
+    public static Vector2I<TScalar> One { get; }
+
+    public TScalar this[int index] { get; }
+
+    public Vector2I(TScalar x, TScalar y);
```

Comment:
Missing:

`Vector2I(TScalar value)` repeating value constructor
`Vector2I(ReadOnlySpan<TScalar> span)` span constructor

also worth noting for larger vectors that this will also be available:
`Vector3I(Vector2I<TScalar> vector, TScalar z)` etc.

Reply:
Fixing the first two, the third is described in a comment at the beginning of this section.